### PR TITLE
Treat unmatched JSON type during element deserialization of decoding …

### LINKED ISSUE #816: deserialization: IllegalStateException when null occurs in JSON input in place of an expected non-null object
`SerializationException` should be thrown instead.

Example: try to parse this JSON:

```
{"foo":null}
```

for deserializing this class:

```
class MyClass(val foo: SomeObject)
```

This exception is thrown:

```
    java.lang.IllegalStateException: Expected class kotlinx.serialization.json.JsonObject but found class kotlinx.serialization.json.JsonNull
        at kotlinx.serialization.json.internal.AbstractJsonTreeInput.beginStructure(TreeJsonInput.kt:224)
```

### PR DESCRIPTION
**Pull request by qwwdfsad**
…exception instead of ISE to maintain decoding invariant

Fixes #816

### CODE REVIEW

**Comment by qwwdfsad** [comment]
One more native problem, PTAL

**Comment by sandwwraith** [comment]
It's the same one, I think

File: runtime/commonTest/src/kotlinx/serialization/json/polymorphic/JsonPolymorphismExceptionTest.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Test
+    fun testDecodingException() = parametrizedTest { useStreaming ->
+        val serialModule = SerializersModule {
+            polymorphic(Base::class) {
+                subclass<Derived>()
+            }
+        }
+
+        assertFailsWith<JsonDecodingException> {
```
Comment (sandwwraith) [inline runtime/commonTest/src/kotlinx/serialization/json/polymorphic/JsonPolymorphismExceptionTest.kt:32]: Don't you want to check the exception message?

Comment (qwwdfsad) [reply runtime/commonTest/src/kotlinx/serialization/json/polymorphic/JsonPolymorphismExceptionTest.kt:32]: I usually avoid it

**Pull request merged**